storage/src/main/java/org/apache/kafka/storage/internals/checkpoint/PartitionMetadataFile.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.storage.internals.checkpoint;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.errors.InconsistentTopicIdException;
+import org.apache.kafka.common.errors.KafkaStorageException;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.storage.internals.log.LogDirFailureChannel;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+
+public class PartitionMetadataFile {
+    private static final String PARTITION_METADATA_FILE_NAME = "partition.metadata";
+    static final int CURRENT_VERSION = 0;
+
+    public static File newFile(File dir) {
+        return new File(dir, PARTITION_METADATA_FILE_NAME);
+    }
+
+    private final File file;
+    private final LogDirFailureChannel logDirFailureChannel;
+
+    private final Object lock = new Object();
+    private volatile Optional<Uuid> dirtyTopicIdOpt = Optional.empty();
+
+    public PartitionMetadataFile(
+        final File file,
+        final LogDirFailureChannel logDirFailureChannel
+    ) {
+        this.file = file;
+        this.logDirFailureChannel = logDirFailureChannel;
+    }
+
+    /**
+     * Records the topic ID that will be flushed to disk.
+     */
+    public void record(Uuid topicId) {
+        // Topic IDs should not differ, but we defensively check here to fail earlier in the case that the IDs somehow differ.
+        dirtyTopicIdOpt.ifPresent(dirtyTopicId -> {
+            if (dirtyTopicId != topicId) {
```

Comment:
I don't think this should be a reference comparison. Perhaps a miss in the scala -> java conversion.